File: solutions/cpp/0695.cpp
class Solution {
 public:
  int maxAreaOfIsland(vector<vector<int>>& grid) {
    const int m = grid.size();
    const int n = grid[0].size();

    int ans = 0;
    vector<vector<bool>> visited(m, vector<bool>(n));

    for (int i = 0; i < m; ++i)
      for (int j = 0; j < n; ++j) ans = max(ans, dfs(grid, i, j, visited));

    return ans;
  }

 private:
  int dfs(vector<vector<int>>& grid, int i, int j,
          vector<vector<bool>>& visited) {
    if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() ||
        visited[i][j] || grid[i][j] == 0)
      return 0;

    visited[i][j] = true;

    return 1 + dfs(grid, i + 1, j, visited) + dfs(grid, i - 1, j, visited) +
           dfs(grid, i, j + 1, visited) + dfs(grid, i, j - 1, visited);
  }
};

File: solutions/python3/0695.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        def dfs(i: int, j: int) -> int:
            if not 0 <= i < len(grid) or not 0 <= j < len(grid[0]) or visited[i][j] or grid[i][j] == 0:
                return 0

            visited[i][j] = True

            return 1 + dfs(i + 1, j) + dfs(i - 1, j) + dfs(i, j + 1) + dfs(i, j - 1)

        m = len(grid)
        n = len(grid[0])

        ans = 0
        visited = [[False] * n for _ in range(m)]

        for i in range(m):
            for j in range(n):
                if not visited[i][j]:
                    ans = max(ans, dfs(i, j))

        return ans
